### ISSUE TITLE
Use with OTP/Elixir, and specifically Phoenix LiveView

### ISSUE THREAD

**Issue by mayel** [tauri-apps/tauri#1270]
**Is your feature request related to a problem? Please describe.**

Glad to see a solid alternative to Electron! While Rust is a great choice (though I am currently building apps in Elixir) I much prefer not using JS-heavy frameworks like React and instead use [Phoenix's LiveView](https://www.phoenixframework.org/) to build SPA apps driven by the backend.

**Describe the solution you'd like**

I'd love see what it would take to bundle OTP and an Elixir and LiveView-based app using Tauri. 

**Describe alternatives you've considered**

[Bakeware](https://github.com/spawnfest/bakeware) is a step in the right direction, and could maybe be combined with Tauri for these purposes.

**Comment by tensor-programming** [comment]
Hmm, it might be interesting to do this.  Really, would just need to use bakeware or elixirscript and then sideload the pheonix/elixir binary with the Tauri app to produce a binary containing both.  The main difference here between a live view style app and something more traditional, is that you have this server in between, feeding the rendered HTML into the Tauri WebView through the phoenix channel.

The other side of this is that you would likely have to use an embeddable database if you planned to still use ecto. You wouldn't be able to package postgres with the phoenix app.

**Comment by mayel** [comment]
Glad to hear it seems straightforward! I just saw [embedding binaries](https://tauri.studio/en/docs/usage/recipes/bundler/sidecar) is already supported, so I guess we'd mostly need a third [app loading mode](https://tauri.studio/en/docs/usage/modes) to load the Elixir app's port. 

[off topic @tensor-programming you may be interested in [this](https://www.youtube.com/watch?v=fPKMmJpAGWc) if you haven't seen t yet]

**Comment by lucasfernog** [comment]
We should probably work to make this possible with a Tauri plugin.

**Comment by nklayman** [comment]
[Here is an example](https://github.com/nklayman/theia/tree/examples/add-tauri/examples/tauri) of a Tauri app that spawns a binary that starts a server and then navigates the webview to that url.

**Comment by Laegel** [comment]
You should also consider looking for an embeddable database to replace Postgres for your project: good news, we have an SQLite plugin!

**Comment by mayel** [comment]
> The other side of this is that you would likely have to use an embeddable database if you planned to still use ecto. You wouldn't be able to package postgres with the phoenix app.

Yeah that was at the back of my mind... Postgres can do silent installs [on Windows](https://www.postgresql.org/download/windows/), and there's [an app bundle](https://postgresapp.com/) for OSX (and Linux users are used to and usually know how to jump through a few more hoops) but I guess users would have to set that up separately?

**Comment by mayel** [comment]
> You should also consider looking for an embeddable database to replace Postgres for your project: good news, we have an SQLite plugin!

While that may suit a lot of projects, some apps use things like JSONB fields, Postgis, or other features that would at the very least require rewrites and at worst loose important functionality (not to mention performance).

**Comment by tensor-programming** [comment]
@mayel yeah, am aware of Nx.  Yet another really cool idea out of the elixir scene.  

I am looking at building a LiveView example right now just to see whats possible and where the potential bottle necks are.  And I wonder what kind of benefits you would get from this kind of setup.

**Comment by mayel** [comment]
@tensor-programming no pressure, but wondering what you discovered if you got around to playing with this?

**Comment by jonericcook** [comment]
+1 on being excited about the possibility of running a phoenix app on a desktop and mobile 🤞 

@tensor-programming here are some projects doing somewhat similar things

And any news on this?

- https://github.com/elixir-desktop
- https://github.com/lumen/lumen
- https://github.com/bake-bake-bake/bakeware
- https://elixirforum.com/t/best-option-to-do-desktop-apps-in-elixir/36835

**Comment by jonericcook** [comment]
another cool project that could be useful? https://github.com/burrito-elixir/burrito

**Comment by gotjoshua** [comment]
very curious to see examples of tauri apps with elixir!

**Comment by heri16** [comment]
Why not mention of rustler? Elixir apps needs no bindings to tauri backend? 

https://github.com/rusterlium/rustler

**Comment by nothingismagick** [comment]
One of the (only?) concerns with rustler Nifs is that you won't be able to handle extreme crashes... 

https://news.ycombinator.com/item?id=31023587

**Comment by nothingismagick** [comment]
A neat use case for this would be to ship an elixir binary as a sidecar and use phoenix channels and websocket listeners in the webview (or just place liveviews in the webview too). I would expect this to literally work right out of the box.

**Comment by heri16** [comment]
No to Typescript. Prefer just Liveview but doesn't seem to be a way to call Taurus API from Liveview?

**Comment by nothingismagick** [comment]
Liveview is really AFAIK SSR HTML passed to Tauri via websockets. So you HAVE to place the liveview into the webview somehow and that means integrating a websocket.

<img width="1019" alt="image" src="https://user-images.githubusercontent.com/35242872/165094318-b98e9aac-d2b9-4e92-af62-ebe5a43c49c0.png">

**Comment by nothingismagick** [comment]
But we haven't experimented with how you would interact with the API in this way, but it is interesting for sure.

**Comment by feng19** [comment]
I make a demo project for this
lookup here: https://github.com/feng19/phx_new_desktop

**Comment by chris838** [comment]
@feng19 thanks for sharing, this repo is super helpful!

I notice you're calling out to rust/tauri from within javascript (i.e. client-side, if this was a web app). Is it possible to interface with tauri from within Elixir? There is a rust API as well as JS, so not sure if this is possible from within a Rust NIF in Elixir.

**Comment by feng19** [comment]
> @feng19 thanks for sharing, this repo is super helpful!
> 
> I notice you're calling out to rust/tauri from within javascript (i.e. client-side, if this was a web app). Is it possible to interface with tauri from within Elixir? There is a rust API as well as JS, so not sure if this is possible from within a Rust NIF in Elixir.

Erlang VM and Tauri are separate processes, there is no ready-made interface, maybe you can have a try.

**Comment by Exadra37** [comment]
I am working in [this fork](https://github.com/Exadra37/ex_tauri) to improve [ex_tauri](https://github.com/filipecabaco/ex_tauri) from @filipecabaco , but the most challenging part is not being able to cross compile from Linux to other platfroms, and not be able to use Tauri actions to compile an actual release that works on Linux, but I can compile on locally that runs fine :). Eventually some day I will manage to get this to work ;)

**Comment by tensor-programming** [comment]
Ill come back to this at some point, just got really busy.    It is possible to do it with Nifs, I was able to at least confirm that.  My main blocker is whether or not its actually worth it given how much you have to pack into the tauri app to make the liveview work.

**Comment by filipecabaco** [comment]
The one we've been using for ex_tauri as actually worked great, biggest issue is moving to manifest 2.0 of Tauri, streamline workflows and have a channel of communication between Rust (Tauri) and the sidecar (Elixir) which I want to use websocket to keep it extremely simple.

here's a talk I've done aong time ago: https://youtu.be/xwdJsAIpxEE?si=Cf0Dy9mFqfm1ktIx

**Comment by feng19** [comment]
A new demo for this: https://github.com/feng19/x_trace_desktop

- upgrade to tauri 2.0
- Internal address: http://localhost:port

> I make a demo project for this lookup here: https://github.com/feng19/phx_new_desktop

**Comment by UtkarshUsername** [comment]
Hi @mayel @tensor-programming! Any update on this?

Here is something that might be useful:
https://mrpopov.com/posts/elixir-liveview-single-binary/

**Comment by filipecabaco** [comment]
I'm trying to finish the https://github.com/filipecabaco/ex_tauri project as I want some helpers on the Elixir side to communicate with Rust so we can send notifications for example. Hopefully I can move forward with this to a RC state by end of Feb

**Comment by UtkarshUsername** [comment]
Awesome work @filipecabaco. I will try using it

**Comment by filipecabaco** [comment]
Please do and do provide feedback. Some of the things feel ok but I really want to have more features and streamline the experience
